lightning/src/chain/chainmonitor.rs

```diff
@@ -152,31 +255,76 @@ where C::Target: chain::Filter,
 	/// calls must not exclude any transactions matching the new outputs nor any in-block
 	/// descendants of such transactions. It is not necessary to re-fetch the block to obtain
 	/// updated `txdata`.
-	fn process_chain_data<FN>(&self, header: &BlockHeader, txdata: &TransactionData, process: FN)
+	///
+	/// Calls which represent a new blockchain tip height should set `best_height`.
+	fn process_chain_data<FN>(&self, header: &BlockHeader, best_height: Option<u32>, txdata: &TransactionData, process: FN)
 	where
 		FN: Fn(&ChannelMonitor<ChannelSigner>, &TransactionData) -> Vec<TransactionOutputs>
 	{
 		let mut dependent_txdata = Vec::new();
-		let monitor_states = self.monitors.read().unwrap();
-		for monitor_state in monitor_states.values() {
-			let mut txn_outputs = process(&monitor_state.monitor, txdata);
+		if let Some(height) = best_height {
+			// Sadly AtomicUsize::fetch_max wasn't stabilized until 1.45, so until then we have to
+			// manually CAS.
+			let mut old_height = self.highest_chain_height.load(Ordering::Relaxed);
+			let new_height = height as usize;
+			while new_height > old_height && self.highest_chain_height
+				.compare_exchange(old_height, new_height, Ordering::AcqRel, Ordering::Relaxed).is_err()
+			{
+				old_height = self.highest_chain_height.load(Ordering::Acquire);
+			}
```

Comment:
I'd find this simpler/more readable: 

```
let old_height = ..
let new_height = ..
if new_height > old_height {
    self.highest_chain_height.store(new_height)
}
```

maybe I'm missing something here?

Reply:
Duh, yes, can do that as long as its under the lock. I've moved the code down and simplified.